File: sampleoutput.py
#Importing the libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
% matplotlib inline
import math
import os
from sklearn import svm,datasets
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency
from sklearn.cross_validation import train_test_split
from sklearn import tree
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, \
    recall_score, confusion_matrix, classification_report, \
    accuracy_score, f1_score
from sklearn import neighbors, datasets
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics

#Importing the dataset
os.chdir('C:/Users/chait/Desktop/Project1')
df = pd.read_csv('Train_data.csv')

#Data Preprocessing
df.dtypes
df.isnull().sum()
df['area code'] = df['area code'].astype(object)
str(df['area code'])
df.info()
df.drop(['phone number'],axis = 1,inplace = True)

#Dividing the numeric and categorical variables
cnames = df.select_dtypes(exclude = ['object'])
numeric_variables = list(cnames.columns.values)

#Saving Categoical columns
cat_data = df.select_dtypes(include = ['object'])
categorical_variables = list(cat_data.columns.values)

#Assigning levels to the categories
for i in range(0,df.shape[1]):
    if(df.iloc[:,i].dtypes == 'object'):
        df.iloc[:,i] = pd.Categorical(df.iloc[:,i])
        df.iloc[:,i] = df.iloc[:,i].cat.codes
for i in categorical_variables:
    df[i] = df[i].astype(object)

#Exploratory Data Analysis
sns.countplot(x="Churn",data=df)

#Outlier Analysis
plt.boxplot(df['account length'])
for i in numeric_variables:
    print(i)
    q75,q25 = np.percentile(df.loc[:,i],[75,25])
    iqr = q75-q25
    
    min = q25 - (iqr*1.5)
    max = q75 + (iqr*1.5)
    df.loc[df.loc[:,i]<min ,i] = np.nan
    df.loc[df.loc[:,i]>max ,i] = np.nan

#Replacing the outliers with median value    
df = df.fillna(df.median(),inplace = True)

for i in categorical_variables:
    df[i] = df[i].astype(object)

cnames = df.select_dtypes(exclude = ['object'])
numeric_variables = list(cnames.columns.values)

#Feature Selection
f,ax = plt.subplots(figsize =(7,5))

#Generate correlation matrix
corr = cnames.corr()

#Plot using seaborne library
sns.heatmap(corr, mask =np.zeros_like(corr, dtype = np.bool), 
            cmap =sns.diverging_palette(220, 10, as_cmap =True),
            square = True, ax = ax)
df.drop(['total day minutes','total eve minutes','total night minutes'],axis = 1,
        inplace = True)

cnames = df.select_dtypes(exclude = ['object'])
numeric_variables = list(cnames.columns.values)

#Feature scaling/Normalization
%matplotlib inline
for i in numeric_variables:
    plt.hist(df[i], bins ='auto')
    plt.show

#Normalization of the data
for i in numeric_variables:
    print(i)
    a = df[i].max()
    b = df[i].min()
    df[i] = (df[i]- b)/(a-b)
#Chi-sqare test
categorical_variables = categorical_variables[:-1] 
for i in categorical_variables:
    print(i)
    chi2, p, dof, ex =chi2_contingency(pd.crosstab(df.Churn,df[i]))
    print(p)
df.drop(['area code','voice mail plan'],axis = 1, inplace =True)
X_train = df.iloc[:,0:14]
Y_train = df.iloc[:,14]
Y_train = Y_train.astype('int') 

#########Importing the test dataset#########

df = pd.read_csv('Test_data.csv')

#Preparing the sample input data
df = df.sample(n = 500)

#Data Preprocessing
df.isnull().sum()
#df['area code'] = df['area code'].astype(object)
df.drop(['phone number'],axis = 1,inplace = True)

#Dividing the numeric and categorical variables
cnames = df.select_dtypes(exclude = ['object'])
numeric_variables = list(cnames.columns.values)

#Saving Categoical columns
cat_data = df.select_dtypes(include = ['object'])
categorical_variables = list(cat_data.columns.values)

#Assigning levels to the categories
for i in range(0,df.shape[1]):
    if(df.iloc[:,i].dtypes == 'object'):
        df.iloc[:,i] = pd.Categorical(df.iloc[:,i])
        df.iloc[:,i] = df.iloc[:,i].cat.codes
for i in categorical_variables:
    df[i] = df[i].astype(object)

#Feature Selection
df.drop(['total day minutes','total eve minutes','total night minutes'],axis = 1,
        inplace = True)
df.drop(['area code','voice mail plan'],axis = 1, inplace =True)

cnames = df.select_dtypes(exclude = ['object'])
numeric_variables = list(cnames.columns.values)

#Saving Categoical columns
cat_data = df.select_dtypes(include = ['object'])
categorical_variables = list(cat_data.columns.values)

#Normalization of the data
for i in numeric_variables:
    print(i)
    a = df[i].max()
    b = df[i].min()
    df[i] = (df[i]- b)/(a-b)

X_test = df.iloc[:,0:14]
Y_test = df.iloc[:,14]

#Converting the target variable to int type for
Y_test = Y_test.astype('int') 
Y_train = Y_train.astype('int')
 
#Decision Tree
clf = tree.DecisionTreeClassifier(criterion ='entropy').fit(X_train,Y_train)

#Predict
Y_pred = clf.predict(X_test)
fpr, tpr, threshold = metrics.roc_curve(Y_test, Y_pred)

#Check accuracy of the model
print ('Accuracy:', accuracy_score(Y_test, Y_pred)*100)
print ('F1 score:', f1_score(Y_test, Y_pred))
print ('Sensitivity or Recall:', recall_score(Y_test, Y_pred))
print ('Precision:', precision_score(Y_test, Y_pred))
print ('\n clasification report:\n', classification_report(Y_test, Y_pred))
print ('\n confussion matrix:\n',confusion_matrix(Y_test, Y_pred))

#ROC PLot
# calculate the fpr and tpr for all thresholds of the classification

fpr, tpr, threshold = metrics.roc_curve(Y_test, Y_pred)
roc_auc = metrics.auc(fpr, tpr)

plt.title('Receiver Operating Characteristic')
plt.plot(fpr, tpr, 'b', label = 'AUC = %0.2f' % roc_auc)
plt.legend(loc = 'lower right')
plt.plot([0, 1], [0, 1],'r--')
plt.xlim([0, 1])
plt.ylim([0, 1])
plt.ylabel('True Positive Rate')
plt.xlabel('False Positive Rate')
plt.show()

df.to_csv('sampleinput_python.csv')
np.savetxt("output_python.csv", Y_pred, newline=" ")
########################END###########################

File: sampleoutput.R
rm(list = ls())

#Load Libraries
x = c('ggplot2', 'corrgram', 'DMwR', 'caret', 'randomForest', 'unbalanced', 'C50', 'dummies', 
      'e1071', 'Information','MASS','rpart', 'gbm', 'ROSE') 

#install.packages(x)
lapply(x, require, character.only = TRUE)
rm(x)
library(ggplot2)
#Importing the dataset

setwd('C:/Users/chait/Desktop/Project1')
df = read.csv('Train_Data.csv')

#Data preprocessing
sum(is.na(df))
head(df)
str(df)
df$state = as.factor(as.character(df$state))
df$area.code = as.factor(as.character(df$area.code))
str(df$state)
str(df$area.code)
df = subset(df,select =-c(phone.number))

#Exploratory data analysis
library(reshape)
table(df$Churn)
pie(table(df$Churn),main = 'Pie chart for Customer Churn',radius = 1)

#Exploratory data analysis
imppred <- randomForest(Churn ~ ., data = df,
                        ntree = 100, keep.forest = FALSE, importance = TRUE)
importance(imppred, type = 1)

#Seperating the numeric data
numeric_index = sapply(df,is.numeric)
numeric_data = df[,numeric_index]

cnames = colnames(numeric_data)

#Encoding categorical variables 
for(i in 1:ncol(df)){
  if(class(df[,i]) == 'factor'){
    df[,i] = factor(df[,i], labels =(1:length(levels(factor(df[,i])))))
  }
}

#Outlier analysis
for (i in 1:length(cnames)){
  assign(paste0("gn",i),ggplot(aes_string(y = cnames[i],x = "Churn"),data = subset(df))+
           stat_boxplot(geom = "errorbar", width = 0.5)+
           geom_boxplot(outlier.colour = 'red', fill = "green", outlier.shape = 18,
                        outlier.size = 1, notch = FALSE) +
           theme(legend.position = "bottom")+
           labs(y=cnames[i],x = "Churn")+
           ggtitle(paste(" ", cnames[i])))
}
#Plotting plots together
gridExtra::grid.arrange(gn1,gn2,gn3,ncol = 3)
gridExtra::grid.arrange(gn4,gn5,gn6,ncol = 3)
gridExtra::grid.arrange(gn7,gn8,gn9,ncol = 3)
gridExtra::grid.arrange(gn10,gn11,gn12,ncol = 3)
gridExtra::grid.arrange(gn13,gn14,gn15,ncol = 3)

#Imputing the outlier values with knnImputation method
for( i in cnames){
  print(i)
  val = df[,i][df[,i] %in% boxplot.stats(df[,i])$out]
  df[,i][df[,i] %in% val] = NA
}

df = knnImputation(df, k=80)

#Feature Selection | Correlation plot for numeric variables
library(corrplot)

m = cor(df[,numeric_index])
corrplot(m, method="square")

#Feature Selection 
factor_index = sapply(df,is.factor)
factor_data = df[,factor_index]

for(i in 1:4)
{
  print(names(factor_data[i]))
  print(chisq.test(table(factor_data$Churn,factor_data[,i])))
  
}

#Feature Selection or Dimensionality Reduction
df1 = subset(df, select = -c(total.day.charge,total.eve.charge,total.night.charge,total.intl.charge,area.code,voice.mail.plan))

numeric_index = sapply(df1,is.numeric)
numeric_data = df1[,numeric_index]

df_train=df1

#Feature Scaling | Normalizing the data
for( i in colnames(numeric_data))
{
  print(i)
  df_train[,i] = (df_train[,i]- min(df_train[,i]))/
    (max(df_train[,i] - min(df_train[,i])))   
}

############ Test Data processing############

#Importing the test data
df = read.csv('Test_data.csv')
df=df[sample(nrow(df), 500), ]
write.csv(df,'sampleinput_R.csv')
sum(is.na(df))
head(df)
str(df)

#Basic pre-processing
df$state = as.factor(as.character(df$state))
df$area.code = as.factor(as.character(df$area.code))
str(df$area.code)
datatypes = sapply(df,class)
for(i in 1:ncol(df)){
  if(class(df[,i]) == 'factor'){
    df[,i] = factor(df[,i], labels =(1:length(levels(factor(df[,i])))))
  }
}

#Feature Selection or Dimensionality Reduction in Test Data
df1 = subset(df, select = -c(total.day.charge,total.eve.charge,total.night.charge,total.intl.charge,phone.number,area.code,voice.mail.plan))

df_test=df1

#Feature scaling | Normalization
for( i in colnames(numeric_data))
{
  print(i)
  df_test[,i] = (df_test[,i]- min(df_test[,i]))/
    (max(df_test[,i] - min(df_test[,i])))   
}

#Building and training a decision tree model
#install.packages('C50')
library(C50)
C50_model = C5.0(Churn ~.,df_train, trails = 100, rules = TRUE)

#summary(C50_model)
#Predicting the output
C50_predictions = predict(C50_model,df_test[,-15],type = 'class')
write.csv(C50_predictions,'output_R.csv')

#Evaluate the performance of classification model
Confmatrix_C50 = table(df_test$Churn,C50_predictions)
confusionMatrix(Confmatrix_C50)

#############################################END########################  
